### DISCUSSION TITLE
How do I read/write from/to a dynamic AWS S3 location in a sanitized query (Python Client)?

### DISCUSSION THREAD

**Discussion by qdegraaf** [duckdb/duckdb#10278] [Q&A]
### Duplicate?

I found a similar unanswered question under General: https://github.com/duckdb/duckdb/discussions/10100 . Unsure if this counts as a duplicate as it's under a different category and slightly different (i.e. "if the question under https://github.com/duckdb/duckdb/discussions/10100 is true what would be the best way to approach the scenario listed below?"). Let me know if it is and I will remove this and add my examples to the General thread.

### Question

I am rewriting some data flows to use DuckDB. One of the steps for these flows is to read data from AWS S3, filter it, then write to a different S3 location. The queries would look something like:
```python
    con.execute("""
    CREATE TEMPORARY TABLE temp AS
    SELECT * FROM read_csv('s3://input-bucket/foo/bar/input.csv.gz', AUTO_DETECT=TRUE)
    WHERE some_column = $1
    """, ["filter_value"])
    con.execute("COPY temp TO 's3://target-bucket/bla/output.csv'")
```
The above works perfectly. However the input and output paths for these flows are often dynamic, based on, for example, the date the flow is run. I cannot get the prepared statement to work with S3 paths unfortunately. Some things I have tried below with the errors they give me (similar examples for the input part also did not work)
```python
# Error: Invalid Input Error: Prepared statement needs 0 parameters, 1 given
con.execute("COPY temp TO 's3://target-bucket/bla/%s.csv'", ["output_file_name"])
# Error: Invalid Input Error: Parameter argument/count mismatch, identifiers of the excess parameters: foo
con.execute("COPY temp TO 's3://target-bucket/bla/$foo.csv'", {"foo": "output_file_name"})
# duckdb.duckdb.ParserException: Parser Error: syntax error at or near "||"
# LINE 1: COPY temp TO 's3://target-bucket/bla/' || $1 || '.csv'
con.execute("COPY temp TO 's3://target-bucket/bla/' || $1 || '.csv'", ["output_file_name"])
# duckdb.duckdb.ParserException: Parser Error: syntax error at or near "SELECT"
# LINE 3:         COPY temp TO SELECT (SELECT path FROM output_path)
con.execute("""
    CREATE TEMPORARY TABLE output_path AS SELECT format('s3://target-bucket/{}.csv', $1)::VARCHAR as path;
    COPY temp TO SELECT (SELECT path FROM output_path)
    """, ["output_file_name"])
```
It appears DuckDB (or the Python client) only wants to accept a pure string as a value for `read_csv()` or target of a `COPY TO` statement. Formatting the query without a prepared statement would be vulnerable to injection. Aside from this DuckDB works great so I was wondering:

- Is there any way to combine a prepared statement with a dynamic S3 path in DuckDB?
- If not, does DuckDB provide other means of minimising the risk of SQL Injection (aside from custom SQL sanitising logic before interpolating the query string)?

Thanks in advance

**Comment by renzepost** [comment]
I have a very similar situation. Just to give an example of what I'd like to see, in `psycopg2` you can do something like this:

```python
table = psycopg2.sql.Identifier("myschema", "mytable")
sql = psycopg2.sql.SQL("SELECT * FROM {}").format(table)
```
But in DuckDB, there doesn't seem to be a way to parameterize identifiers. The only solution seems to be to concatenate the SQL string, exposing yourself to SQL injection.

Edit: I did some googling and I understand it's not very common to be able to parameterize/sanitize identifiers such as table names in prepared statements (in other db systems). So I guess string concatenation is the only option, and you'd have to make sure that it doesn't contain any user input. In my case, I want to attach to a Postgres database in Apache Airflow, and I'm now doing this:

```python
con.execute(f"ATTACH '{postgres_conn.get_uri()}' AS db (TYPE postgres)")
```
I guess within this context the risk of SQL injection is low.
